### ISSUE TITLE
Consider dropping support for format and title in axis and legends for Vega Lite 3.0

### ISSUE THREAD

**Issue by Saba9** [vega/vega-lite#4057]
Proposed by @domoritz. Allowing users to to define format and title in two locations seems confusing. 
One example of potentially confusing behavior is that when generating tool tips we use the encoding title and format and not the axis title and format which might not be straight forward for users.

**Comment by kanitw** [comment]
FYI, currently there is only `format` for `text`, but not other channels. So I don't know if we want a breaking changes there, especially that the format are only for axis and legend labels, but have no effect on the data fields itself. 

For `title`, it will be also weird if there are properties for enabling parts of axis ( `ticks`, `labels`, `domain`, `grid`), but no `title` property.  So if we gonna change, a sensible change would be changing title to be boolean (so it's more consistent with `ticks`, `labels`, `grid`, `domain`).

**Comment by domoritz** [comment]
I'm not so sure about positioning titles that way but do you agree that we should change `title` to only support booleans?

**Comment by domoritz** [comment]
Any objections @arvind @jheer?

**Comment by arvind** [comment]
Makes sense to me!

**Comment by kanitw** [comment]
I don't know if we should really introduce this breaking changes given it's probably already widely used.  In fact, we even use `axis.title` in the figures for Making Data Visual. 

While we can of course go update the online gallery for making data visual, it seems like we don't gain much from this breaking change -- maybe a better thing to do is fix the bug in tooltip encoding to read axis/legend/header title too?

**Comment by domoritz** [comment]
I think it's better to have only one way to set the title, rather than four (title, axis title, legend title, header title). We only used `axis.title` because that was the only way. Yes, it's a breaking change but we could add a simple preprocessor that copies `axis.title` to `title` when it's not just a boolean. I think now is a good time to change this.

**Comment by kanitw** [comment]
> Yes, it's a breaking change but we could add a simple preprocessor that copies axis.title to title when it's not just a boolean. I think now is a good time to change this.

If you're adding a pre-processor, then it's not changing the behavior then.  Just how we implement it.

**Comment by domoritz** [comment]
Well, and I'd not document it officially. Just to not annoy users too much.

**Comment by kanitw** [comment]
> Well, and I'd not document it officially. Just to not annoy users too much.

I don't see how it's more annoying for users.  Having document that doesn't match behavior is weird. (I think having two ways of doing things is actually more annoying for us who maintain the projects.) 

Thinking more, it's also kinda weird that they can specify `title` outside but then disable it in `axis`.  In a way, if they don't want a title in a first place, they could just specify title = '' from the outside.  

If we start from scratch, I can see the argument for having only one of this.  But at this point, removing it isn't really beneficial and will only break things and annoy users.  So let's only discuss how we gonna refactor this internal and prevent future errors like we have with tooltip. 

I think we should think about fieldDef.title as a shorthand and axis/legend/header.title as the full form (as we will merge `axis/legend/header.title` in our composition logic).  
Basically, `encoding` on full spec can have `fieldDef.title`, but internal encoding shouldn't. 
Otherwise, it's kinda weird that we normalize `axis/legend/header.title` to `fieldDef.title` only to later copy `fieldDef.title` back to axis/legend/header.title` inside.

**Comment by domoritz** [comment]
> Thinking more, it's also kinda weird that they can specify title outside but then disable it in axis. In a way, if they don't want a title in a first place, they could just specify title = '' from the outside.

No, you could have a title that is used for tooltips but not for the axis. Or it's only visible if you use the channel for a legend.

**Comment by kanitw** [comment]
> Or it's only visible if you use the channel for a legend.

I agree with this part. 

> No, you could have a title that is used for tooltips but not for the axis.

Not sure how important is this case, I can see your point. 
If we want to support this, we have two options:

1) normalize to `fieldDef.title` (but it will be weird like I said)
or 2) just fix the bug for tooltip 

I think 2) is way less work and doesn't make the code more complicated. (It's complicated, but I don't see 1) being simpler.)

**Comment by kanitw** [comment]
We fixed the tooltip bug in #4076 and we agreed in person that we should not change the syntax, so let's close this.
